centerofci/mathesar
# Shared queries - Auth handling for query requests, frontend consumer view, basic API tests

Fixes #3034 
Fixes #3036 

*The diff appears large due to the tests added, the actual code is of reasonable size.*
*This is mainly frontend code, the backend changes are minimal.*

**This PR**:
* Bypasses authentication for query requests for valid share uuids
* Adds custom common_data for shared tables required by frontend
* Allows viewing shared queries on the frontend
* Adds basic share crud API tests

**Flows built in this PR:**
* Requests to the following endpoints should succeed with a valid `share-link-uuid`:
   - `/api/db/v0/queries/<query_pk/?shared-link-uuid=<uuid>`
   - `/api/db/v0/queries/<query_pk/results/?shared-link-uuid=<uuid>`
* The following user flows should work for the above endpoints:
   - Unauthenticated requests with valid `shared-link-uuid`
   - Authenticated requests made by users who do not have explicit access the query, but have valid `shared-link-uuid`
   - Authenticated requests made by users who have access to the query, with or without valid `shared-link-uuid`

**Testing instructions:**
* Generate a shared link for a query via API:
  - A post request to `/api/ui/v0/queries/1/shares` with `{ enabled: true }` should share the query and generate a slug. The response will be like follows:
    ```
    {
      "id": 1,
      "slug": "584cfd0p-890j-95ec-9ded-c226d46f9542",
      "enabled": true
    }
    ```
* Visit `http://localhost:8000/shares/explorations/584cfd0p-890j-95ec-9ded-c226d46f9542/` and you should be able to see a readonly view of the query.
* The following cases should be taken into account, when the link is valid:
  - a. The link should work when you're not logged into Mathesar.
  - b. The link should work when the user is logged in but does not have access to the query's schema or DB.
  - c. The link should work when the user is logged in and has access to the table.

**Frontend notes**:
* The FE initially used only the `/run` endpoint for running queries to keep code common between exploration viewing route, and data explorer routes.
* However, permission checks made it clear that the FE had to use the `/results` endpoint for the exploration page, which is also the proper endpoint to use for it. We didn't do this earlier due to a time crunch. This change is implemented in this PR.

**Backend notes**:
* The tests added are only for CRUD operations on /shares endpoints of tables and queries.
* The auth bypass tests for the /tables/* and queries/* endpoints are not implemented in this PR.
* They will be covered in upcoming PRs. 

## Checklist
 
  index.md`).
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
